tangelo/linq/simulator.py

```diff
@@ -26,515 +26,35 @@
-                frequencies[self.__int_to_binstr(i, n_qubits)] = frequency
-
-        # If n_shots, has been specified, then draw that amount of samples from the distribution
-        # and return empirical frequencies instead. Otherwise, return the exact frequencies
-        if not self.n_shots:
-            return frequencies
-        else:
-            xk, pk = [], []
-            for k, v in frequencies.items():
-                xk.append(int(k[::-1], 2))
-                pk.append(frequencies[k])
-            distr = stats.rv_discrete(name='distr', values=(np.array(xk), np.array(pk)))
-
-            # Generate samples from distribution. Cut in chunks to ensure samples fit in memory, gradually accumulate
-            chunk_size = 10**7
-            n_chunks = self.n_shots // chunk_size
-            freqs_shots = Counter()
-
-            for i in range(n_chunks+1):
-                this_chunk = self.n_shots % chunk_size if i == n_chunks else chunk_size
-                samples = distr.rvs(size=this_chunk)
-                freqs_shots += Counter(samples)
-            freqs_shots = {self.__int_to_binstr_lsq(k, n_qubits): v / self.n_shots for k, v in freqs_shots.items()}
-            return freqs_shots
 
-    def __int_to_binstr(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the order
-        specified for the state vector.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1] if (self.statevector_order == "msq_first") else state_binstr
 
-    def __int_to_binstr_lsq(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the
-        least-significant qubit order.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1]
+def Simulator(target=default_simulator, n_shots=None, noise_model=None, **kwargs) -> SimulatorBase:
+    """Return requested target simulator
+
+    Args:
+        target (string or BaseSimulator): String can be "qiskit", "cirq", "qdk" or "qulacs". Can also provide
+            The child class of BaseSimulator.
+        n_shots (int): Number of shots if using a shot-based simulator.
+        noise_model: A noise model object assumed to be in the format
+            expected from the target backend.
+        kwargs: Other arguments that could be passed to a target. Examples are qubits_to_use for a QPU, transpiler optimization
+            level, error mitigation flags etc.
+
+    Returns:
+        BaseSimulator: The initialized target simulator that is a child class of BaseSimulator.
+    """
+    if target is None:
+        target = default_simulator
+    if isinstance(target, str):
+        simulator = target_dict[target](n_shots=n_shots, noise_model=noise_model)
+    else:
+        simulator = target(n_shots=n_shots, noise_model=noise_model, **kwargs)
```

Comment:
What is the purpose of this branch ? (this suggests that regardless of your answer, we need comments).
If its purpose is redundant with the other branch, I suggest we remove it.

If its purpose brings out a new functionality / use cases, I need this functionality to be explained and tested. Unless I missed something, I don't think this has been tested. You have the option of leaving it out for another small PR, this way Simulator and Translator 2.0 can be merged a bit faster first.

Reply:
This was tested in a previous version. It can be seen in (https://github.com/JamesB-1qbit/Tangelo/blob/refactor_simulator_noisy/examples/linq/qiskit_device.ipynb) where QiskitDevice is provided.

Added test for providing user provided subclass of SimulatorBase target.

The branch was needed to allow users to provide their own subclass of SimulatorBase. This is necessary such that algorithms such as VQESolver can be provided with the class TargetSimulator instead of "target" string.for the backend options.

Reply:
Could we add a test that demonstrate that indeed an existing algorithm can work with a custom backend ? Even if the simulator is a dummy and the result somewhat non-sensical, et predictable (e.g maybe the simulator returns a hardcoded statevector, maybe the zero state, which makes the optimization process trivial and yield a particular energy)